File: plugins/tokenmonitor/server/js/src/updatecheck.js
// updatecheck answers one question the broker cannot otherwise see: is a newer
// TokenMonitor broker/plugin release published than the one this process is
// running? The broker does NOT auto-update, so over time it drifts behind the
// firmware it feeds. This module periodically fetches the public marketplace
// catalog, compares the tokenmonitor entry's version against the installed
// release version, and stashes the verdict in the shared State so three
// surfaces can advertise it: the /device/<id>/sync body (-> on-device banner),
// tokenmonitor_health / tokenmonitor_status (-> Claude Code), and a stderr WARN.
//
// It is strictly best-effort: any network/parse failure leaves the cached
// verdict `known:false` (never a false "up to date" or "outdated") and never
// blocks or errors the broker. Wire-identical to the Go internal/updatecheck.

import { readFileSync } from "node:fs";
import { join } from "node:path";
import process from "node:process";

import { compareSemver } from "./ota.js";
import { VERSION } from "./version.js";

// PluginName is the marketplace entry whose version tracks releases.
export const PLUGIN_NAME = "tokenmonitor";

// DEFAULT_MARKETPLACE_URL is the raw catalog on the marketplace repo's default
// branch — the single source of truth for "latest published". Overridable via
// TOKENMONITOR_MARKETPLACE_URL (used by tests).
export const DEFAULT_MARKETPLACE_URL =
  "https://raw.githubusercontent.com/fractal-manifold/mcp-marketplace/main/.claude-plugin/marketplace.json";

const HTTP_TIMEOUT_MS = 10_000;
const POLL_INTERVAL_MS = 6 * 60 * 60 * 1000; // 6h
const INITIAL_DELAY_MS = 30_000; // 30s
const MAX_BODY = 1 * 1024 * 1024;

// marketplaceURL returns the catalog URL, honouring the test/CI override.
export function marketplaceURL() {
  return process.env.TOKENMONITOR_MARKETPLACE_URL || DEFAULT_MARKETPLACE_URL;
}

// installedVersion resolves the running release version. It prefers the
// bundle's plugin.json (the release/marketplace axis, apples-to-apples with the
// catalog) found via CLAUDE_PLUGIN_ROOT, and falls back to the baked-in broker
// build VERSION when that file is absent or unreadable. Mirrors Go
// InstalledVersion.
export function installedVersion(baked = VERSION) {
  const root = process.env.CLAUDE_PLUGIN_ROOT;
  if (root) {
    try {
      const raw = readFileSync(join(root, ".claude-plugin", "plugin.json"), "utf8");
      const m = JSON.parse(raw);
      if (m && typeof m.version === "string" && m.version) return m.version;
    } catch {
      // fall through to baked
    }
  }
  return baked;
}

// fetchLatest GETs the marketplace catalog and returns the tokenmonitor entry's
// version. An empty string means the entry was absent; throws on any HTTP/parse
// error. Callers treat both as "advertise nothing".
export async function fetchLatest(url = marketplaceURL()) {
  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), HTTP_TIMEOUT_MS);
  let resp;
  try {
    resp = await fetch(url, {
      method: "GET",
      redirect: "follow",
      signal: ctrl.signal,
      headers: { Accept: "application/json", "User-Agent": "tokenmonitor-mcp-updatecheck" },
    });
  } finally {
    clearTimeout(timer);
  }
  if (!resp.ok) throw new Error(`marketplace fetch: HTTP ${resp.status}`);
  const buf = Buffer.from(await resp.arrayBuffer());
  const doc = JSON.parse(buf.subarray(0, MAX_BODY).toString("utf8"));
  const plugins = Array.isArray(doc?.plugins) ? doc.plugins : [];
  for (const p of plugins) {
    if (p && p.name === PLUGIN_NAME) return String(p.version || "");
  }
  return "";
}

// check performs one fetch+compare and returns the verdict. On any failure it
// returns { known:false, current } — callers must treat that as "advertise
// nothing". Mirrors Go Check.
export async function check(current, url = marketplaceURL()) {
  let latest;
  try {
    latest = await fetchLatest(url);
  } catch {
    return { known: false, outdated: false, current, latest: "", checkedAt: 0 };
  }
  if (!latest) return { known: false, outdated: false, current, latest: "", checkedAt: 0 };
  const cmp = compareSemver(latest, current);
  if (cmp === null) {
    // Either version is unparseable under the project's semver subset; don't
    // guess.
    return { known: false, outdated: false, current, latest: "", checkedAt: 0 };
  }
  return {
    known: true,
    outdated: cmp > 0,
    current,
    latest,
    checkedAt: Math.floor(Date.now() / 1000),
  };
}

// run polls the marketplace catalog on a slow cadence and publishes each
// verdict into `state`. It is best-effort and never throws. Pass an
// AbortSignal to stop it. `baked` is the compiled-in broker version, used only
// as the installed-version fallback. Mirrors Go Run.
export function run(state, { baked = VERSION, logger = null, abortSignal = null } = {}) {
  if (!state) return;
  const current = installedVersion(baked);
  let stopped = false;
  let timer = null;
  const stop = () => {
    stopped = true;
    if (timer) clearTimeout(timer);
  };
  if (abortSignal) {
    if (abortSignal.aborted) return;
    abortSignal.addEventListener("abort", stop, { once: true });
  }
  const tick = async () => {
    if (stopped) return;
    try {
      const info = await check(current);
      state.setUpdate(info);
      if (logger) {
        if (info.known && info.outdated) {
          logger.warn(`updatecheck: WARNING broker ${info.current} is behind published ${info.latest} — update the tokenmonitor plugin`);
        } else if (info.known && logger.info) {
          logger.info(`updatecheck: broker ${info.current} is up to date`);
        }
      }
    } catch {
      // best-effort: never let a check failure escape.
    }
    if (!stopped) {
      timer = setTimeout(tick, POLL_INTERVAL_MS);
      if (timer && typeof timer.unref === "function") timer.unref();
    }
  };
  timer = setTimeout(tick, INITIAL_DELAY_MS);
  // Don't keep the event loop alive for the poller alone.
  if (timer && typeof timer.unref === "function") timer.unref();
}

File: plugins/tokenmonitor/server/go/internal/updatecheck/updatecheck.go
// Package updatecheck answers one question the broker cannot otherwise see:
// is a newer TokenMonitor broker/plugin release published than the one this
// process is running? The broker does NOT auto-update, so over time it drifts
// behind the firmware it feeds. This package periodically fetches the public
// marketplace catalog, compares the tokenmonitor entry's version against the
// installed release version, and stashes the verdict in *state.State so three
// surfaces can advertise it: the /device/<id>/sync body (→ on-device banner),
// tokenmonitor_health / tokenmonitor_status (→ Claude Code), and a stderr WARN.
//
// It is strictly best-effort: any network/parse failure leaves the cached
// verdict Unknown (never a false "up to date" or "outdated") and never blocks
// or errors the broker.
package updatecheck

import (
	"context"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"github.com/fractal-manifold/tokenmonitor-mcp/internal/ota"
	"github.com/fractal-manifold/tokenmonitor-mcp/internal/state"
)

const (
	// PluginName is the marketplace entry whose version tracks releases.
	PluginName = "tokenmonitor"
	// DefaultMarketplaceURL is the raw catalog on the marketplace repo's
	// default branch — the single source of truth for "latest published".
	// Overridable via TMON_MARKETPLACE_URL (used by tests); legacy
	// TOKENMONITOR_MARKETPLACE_URL is still honoured as an alias.
	DefaultMarketplaceURL = "https://raw.githubusercontent.com/fractal-manifold/mcp-marketplace/main/.claude-plugin/marketplace.json"

	httpTimeout  = 10 * time.Second
	pollInterval = 6 * time.Hour
	initialDelay = 30 * time.Second
	maxBody      = 1 * 1024 * 1024
)

// marketplaceDoc is the subset of marketplace.json this package reads.
type marketplaceDoc struct {
	Plugins []struct {
		Name    string `json:"name"`
		Version string `json:"version"`
	} `json:"plugins"`
}

// pluginManifest is the subset of .claude-plugin/plugin.json we read.
type pluginManifest struct {
	Version string `json:"version"`
}

// MarketplaceURL returns the catalog URL, honouring the test/CI override.
// TMON_ is the project's env-var convention; TOKENMONITOR_MARKETPLACE_URL is a
// backward-compat alias (TMON_ wins when both are set).
func MarketplaceURL() string {
	if u := os.Getenv("TMON_MARKETPLACE_URL"); u != "" {
		return u
	}
	if u := os.Getenv("TOKENMONITOR_MARKETPLACE_URL"); u != "" {
		return u
	}
	return DefaultMarketplaceURL
}

// InstalledVersion resolves the running release version. It prefers the
// bundle's plugin.json (the release/marketplace axis, apples-to-apples with the
// catalog). The launcher exports TMON_PLUGIN_ROOT (set on every client, incl.
// Antigravity and the cached Go binary); CLAUDE_PLUGIN_ROOT is the host-provided
// fallback (Claude/Codex only). Falls back to the baked-in broker build version
// when the manifest is absent or unreadable.
func InstalledVersion(baked string) string {
	root := os.Getenv("TMON_PLUGIN_ROOT")
	if root == "" {
		root = os.Getenv("CLAUDE_PLUGIN_ROOT")
	}
	if root != "" {
		p := filepath.Join(root, ".claude-plugin", "plugin.json")
		if b, err := os.ReadFile(p); err == nil {
			var m pluginManifest
			if json.Unmarshal(b, &m) == nil && m.Version != "" {
				return m.Version
			}
		}
	}
	return baked
}

// fetchLatest GETs the marketplace catalog and returns the tokenmonitor entry's
// version. An empty string (nil error) means the entry was absent.
func fetchLatest(ctx context.Context, client *http.Client) (string, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, MarketplaceURL(), nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "tokenmonitor-mcp-updatecheck")
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", &httpError{resp.StatusCode}
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxBody))
	if err != nil {
		return "", err
	}
	var doc marketplaceDoc
	if err := json.Unmarshal(body, &doc); err != nil {
		return "", err
	}
	for _, p := range doc.Plugins {
		if p.Name == PluginName {
			return p.Version, nil
		}
	}
	return "", nil
}

type httpError struct{ code int }

func (e *httpError) Error() string { return "marketplace fetch: HTTP " + itoa(e.code) }

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	var b [12]byte
	i := len(b)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		i--
		b[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

// Check performs one fetch+compare and returns the verdict. On any failure it
// returns a Known:false result (verdict unknown) — callers must treat that as
// "advertise nothing".
func Check(ctx context.Context, client *http.Client, current string) state.UpdateInfo {
	latest, err := fetchLatest(ctx, client)
	if err != nil || latest == "" {
		return state.UpdateInfo{Known: false, Current: current}
	}
	cmp, ok := ota.CompareSemver(latest, current)
	if !ok {
		// Either version is unparseable under the project's semver subset;
		// don't guess.
		return state.UpdateInfo{Known: false, Current: current}
	}
	return state.UpdateInfo{
		Known:     true,
		Outdated:  cmp > 0,
		Current:   current,
		Latest:    latest,
		CheckedAt: time.Now(),
	}
}

// Run polls the marketplace catalog on a slow cadence and publishes each
// verdict into st. It returns when ctx is cancelled. baked is the compiled-in
// broker version, used only as the installed-version fallback.
func Run(ctx context.Context, baked string, st *state.State, logger *log.Logger) {
	if st == nil {
		return
	}
	current := InstalledVersion(baked)
	client := &http.Client{Timeout: httpTimeout}
	timer := time.NewTimer(initialDelay)
	defer timer.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-timer.C:
		}
		info := Check(ctx, client, current)
		st.SetUpdate(info)
		if logger != nil {
			if info.Known && info.Outdated {
				logger.Printf("updatecheck: WARNING broker %s is behind published %s — update the tokenmonitor plugin", info.Current, info.Latest)
			} else if info.Known {
				logger.Printf("updatecheck: broker %s is up to date", info.Current)
			}
		}
		timer.Reset(pollInterval)
	}
}
